### PR TITLE
Integration testing enhancements

### PR DESCRIPTION
**Pull request by yuval-cloudinary**

### CODE REVIEW

File: backend/cloudinary/cloudinary.go [hunk 1]
```diff
@@ -237,20 +364,25 @@ func (f *Fs) List(ctx context.Context, dir string) (fs.DirEntries, error) {
...
+	if f.opt.OptimisticSearch && len(results.Assets) == 0 && retry < 3 {
+		time.Sleep(1 * time.Second)
+		return f.getCLDAsset(ctx, remote, retry+1)
+	}
 	if err != nil || len(results.Assets) == 0 {
 		return nil, fs.ErrorObjectNotFound
 	}
 
-	if len(results.Assets) > 1 {
+	if results.NextCursor != "" {
```
Comment (const-cloudinary) [inline backend/cloudinary/cloudinary.go]: Do we want to remove this?

File: backend/cloudinary/cloudinary.go [hunk 2]
```diff
@@ -66,18 +144,42 @@ type Fs struct {
...
+}
+
+// Mandatory returns whether the option must be parsed or can be ignored
+func (o *updateOptions) Mandatory() bool {
+	return false
+}
+
+// String formats the option into human-readable form
+func (o *updateOptions) String() string {
+	return fmt.Sprintf("updateOptions(%v)", o.UpdateMode)
```
Comment (const-cloudinary) [inline backend/cloudinary/cloudinary.go]: Not sure if this is in use, if yes - please add other fields as well

File: backend/cloudinary/cloudinary.go [hunk 3]
```diff
@@ -237,20 +364,25 @@ func (f *Fs) List(ctx context.Context, dir string) (fs.DirEntries, error) {
 }
 
 // getCLDAsset finds the asset at Cloudinary. If it can't be found it returns the error fs.ErrorObjectNotFound.
-func (f *Fs) getCLDAsset(ctx context.Context, remote string) (*admin.SearchAsset, error) {
+func (f *Fs) getCLDAsset(ctx context.Context, remote string, retry int8) (*admin.SearchAsset, error) {
```
Comment (const-cloudinary) [inline backend/cloudinary/cloudinary.go]: a better name would be something like `tryNum`, it's not necessarily **re**try

File: backend/cloudinary/cloudinary.go [hunk 4]
```diff
@@ -395,16 +548,81 @@ func (o *Object) SetModTime(ctx context.Context, modTime time.Time) error {
...
+				fs.Logf(o, "Unsupported mandatory option: %v", option)
+			}
+		}
 	}
-	return resp.Body, nil
+	if key != "" && value != "" {
+		opts.ExtraHeaders = make(map[string]string)
+		opts.ExtraHeaders[key] = value
+	}
+	for i := 1; i <= 7; i++ {
```
Comment (const-cloudinary) [inline backend/cloudinary/cloudinary.go]: This code looks a bit unclear, maybe add a few comments, explaining what's going on here?

File: backend/cloudinary/cloudinary_test.go [hunk 5]
```diff
@@ -11,8 +11,13 @@ import (
 
 // TestIntegration runs integration tests against the remote
 func TestIntegration(t *testing.T) {
+	name := "TestCloudinary"
 	fstests.Run(t, &fstests.Opt{
-		RemoteName: "TestCloudinary:",
-		NilObject:  (*cloudinary.Object)(nil),
+		RemoteName:      "TestCloudinary:",
```
Comment (const-cloudinary) [inline backend/cloudinary/cloudinary_test.go]: reuse `name` variable here.

**Review by const-cloudinary — CHANGES_REQUESTED** [review]

File: backend/cloudinary/cloudinary.go (hunk 1 above)
Comment (yuval-cloudinary) [reply backend/cloudinary/cloudinary.go]: removing

File: backend/cloudinary/cloudinary.go (hunk 2 above)
Comment (yuval-cloudinary) [reply backend/cloudinary/cloudinary.go]: This is not in use

File: backend/cloudinary/cloudinary.go (hunk 4 above)
Comment (yuval-cloudinary) [reply backend/cloudinary/cloudinary.go]: Added a comment

File: backend/cloudinary/cloudinary_test.go (hunk 5 above)
Comment (yuval-cloudinary) [reply backend/cloudinary/cloudinary_test.go]: Sure

File: backend/cloudinary/cloudinary.go (hunk 2 above)
Comment (const-cloudinary) [reply backend/cloudinary/cloudinary.go]: Can we remove it then?

File: backend/cloudinary/cloudinary.go (hunk 1 above)
Comment (yuval-cloudinary) [reply backend/cloudinary/cloudinary.go]: This is mandatory, similar to the (*Fs) String. I revised the implementation

File: backend/cloudinary/cloudinary.go [hunk 6]
```diff
@@ -239,21 +390,19 @@ func (f *Fs) List(ctx context.Context, dir string) (fs.DirEntries, error) {
...
+			CloudinaryEncoder.FromStandardName(f, path.Base(remote))),
 		MaxResults: 1,
 	}
+	f.WaitEventuallyConsistent()
 	results, err := f.cld.Admin.Search(ctx, searchParams)
 	if err != nil || len(results.Assets) == 0 {
 		return nil, fs.ErrorObjectNotFound
 	}
 
-	if len(results.Assets) > 1 {
```
Comment (const-cloudinary) [inline backend/cloudinary/cloudinary.go:253]: Do we just ignore the rest of the found items?

Comment (yuval-cloudinary) [reply backend/cloudinary/cloudinary.go:253]: That's not the latest code, but basically yes, we are searching with `MaxResults: 1`
This function is called from 
func (f *Fs) NewObject(ctx context.Context, remote string) (fs.Object, error) 
func (f *Fs) Remove(ctx context.Context, o fs.Object) error

I see that for the Remove we can pass the publicId and be specific. For NewObject - there is mechanism to handle duplicate names

Comment (yuval-cloudinary) [reply backend/cloudinary/cloudinary.go:253]: Actually, changing the Remove functionality would not change that much.
What do you expect to happen when someone type:
rclone rm file_with_duplicate_name

**Pull request closed without merging**